// babel-core/lib/babel/traversal/context.js
"use strict";

var _classCallCheck = function (instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } };

var TraversalPath = require("./path");
var flatten = require("lodash/array/flatten");
var compact = require("lodash/array/compact");

var TraversalConext = (function () {
  function TraversalConext(scope, opts, state, parentPath) {
    _classCallCheck(this, TraversalConext);

    this.shouldFlatten = false;
    this.parentPath = parentPath;

    this.scope = scope;
    this.state = state;
    this.opts = opts;
  }

  TraversalConext.prototype.flatten = function flatten() {
    this.shouldFlatten = true;
  };

  TraversalConext.prototype.visitNode = function visitNode(node, obj, key) {
    var iteration = new TraversalPath(this, node, obj, key);
    return iteration.visit();
  };

  TraversalConext.prototype.visit = function visit(node, key) {
    var nodes = node[key];
    if (!nodes) return;

    if (!Array.isArray(nodes)) {
      return this.visitNode(node, node, key);
    }

    // nothing to traverse!
    if (nodes.length === 0) {
      return;
    }

    for (var i = 0; i < nodes.length; i++) {
      if (nodes[i] && this.visitNode(node, nodes, i)) {
        return true;
      }
    }

    if (this.shouldFlatten) {
      node[key] = flatten(node[key]);

      if (key === "body") {
        // we can safely compact this
        node[key] = compact(node[key]);
      }
    }
  };

  return TraversalConext;
})();

module.exports = TraversalConext;
